From 063b2be3bcac70d9d629bfc592f2851492559eb0 Mon Sep 17 00:00:00 2001
From: Aloka Dixit <alokad@codeaurora.org>
Date: Tue, 21 Jul 2020 16:19:43 -0700
Subject: [PATCH] hostapd: Add MBSSID DTIM fields for nontransmitting VAP

This patch adds missing mandatory fields in MBSSID Index element,
DTIM period and count, for non transmitting interfaces.

Known issue: Sometimes DTIM count for one of the profiles shows
	     invalid values, could be in FW.
	     Will be fixed in a different patch if applicable.

Signed-off-by: Aloka Dixit <alokad@codeaurora.org>
---
 src/ap/beacon.c     |  4 ++--
 src/ap/ieee802_11.c | 18 +++++++++++++-----
 src/ap/ieee802_11.h |  3 ++-
 3 files changed, 17 insertions(+), 8 deletions(-)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index b4611494cd4b..8cff43b79306 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -529,7 +529,7 @@ static u8 * hostapd_gen_probe_resp(struct hostapd_data *hapd,
 	pos = hostapd_get_mde(hapd, pos, epos - pos);
 
         if (hapd->iconf->multiple_bssid)
-		pos = hostapd_eid_multiple_bssid(hapd, pos, epos);
+		pos = hostapd_eid_multiple_bssid(hapd, pos, epos, 0);
 
 	/* eCSA IE */
 	csa_pos = hostapd_eid_ecsa(hapd, pos);
@@ -1271,7 +1271,7 @@ int ieee802_11_build_ap_params(struct hostapd_data *hapd,
 	tailpos = hostapd_get_mde(hapd, tailpos, tailend - tailpos);
 
 	if (hapd->iconf->multiple_bssid)
-		tailpos = hostapd_eid_multiple_bssid(hapd, tailpos, tailend);
+		tailpos = hostapd_eid_multiple_bssid(hapd, tailpos, tailend, 1);
 
 	/* eCSA IE */
 	csa_pos = hostapd_eid_ecsa(hapd, tailpos);
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index cfdb1688dce1..dd639880770e 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5686,7 +5686,8 @@ int hostapd_eid_multiple_bssid_len(struct hostapd_data *hapd)
 
 
 static u8 * hostapd_eid_multiple_bssid_chunk(struct hostapd_data *hapd,
-					     u8 *eid, u8 *end, int *count)
+					     u8 *eid, u8 *end, int *count,
+					     u8 is_beacon)
 {
 	u8 *size_offset, *num_offset, num = 0;
 	int i;
@@ -5697,7 +5698,7 @@ static u8 * hostapd_eid_multiple_bssid_chunk(struct hostapd_data *hapd,
 
 	for (i = *count; i < hapd->iface->num_bss; i++) {
 		struct hostapd_data *bss = hapd->iface->bss[i];
-		u8 *bss_size_offset, *pos = eid;
+		u8 *bss_size_offset, *index_size_offset, *pos = eid;
 		u16 capab_info;
 
 		*eid++ = WLAN_EID_SUBELEMENT_NONTRANSMITTED_BSSID_PROFILE;
@@ -5715,8 +5716,13 @@ static u8 * hostapd_eid_multiple_bssid_chunk(struct hostapd_data *hapd,
 		eid += bss->conf->ssid.ssid_len;
 
 		*eid++ = WLAN_EID_MULTIPLE_BSSID_INDEX;
-		*eid++ = 1;
+		index_size_offset = eid++;
 		*eid++ = i;
+		if (is_beacon) {
+			*eid++ = bss->conf->dtim_period;
+			*eid++ = 0xFF;
+		}
+		*index_size_offset = (eid - index_size_offset) - 1;
 
 		eid = hostapd_get_rsne(bss, eid, end - eid);
 		*bss_size_offset = (eid - bss_size_offset) - 1;
@@ -5739,12 +5745,14 @@ multiple_bssid_too_big:
 }
 
 
-u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid, u8 *end)
+u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid, u8 *end,
+				u8 is_beacon)
 {
 	int count = 1;
 
 	while (count < hapd->iface->num_bss)
-		eid = hostapd_eid_multiple_bssid_chunk(hapd, eid, end, &count);
+		eid = hostapd_eid_multiple_bssid_chunk(hapd, eid, end, &count,
+						       is_beacon);
 
 	return eid;
 }
diff --git a/src/ap/ieee802_11.h b/src/ap/ieee802_11.h
index 4ae18e7ee935..35bd06e43883 100644
--- a/src/ap/ieee802_11.h
+++ b/src/ap/ieee802_11.h
@@ -119,7 +119,8 @@ u8 * hostapd_eid_time_zone(struct hostapd_data *hapd, u8 *eid);
 int hostapd_update_time_adv(struct hostapd_data *hapd);
 void hostapd_client_poll_ok(struct hostapd_data *hapd, const u8 *addr);
 u8 * hostapd_eid_bss_max_idle_period(struct hostapd_data *hapd, u8 *eid);
-u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid, u8 *end);
+u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid,
+				u8 *end, u8 is_beacon);
 int hostapd_eid_multiple_bssid_len(struct hostapd_data *hapd);
 
 int auth_sae_init_committed(struct hostapd_data *hapd, struct sta_info *sta);
-- 
2.25.0

